### ISSUE TITLE
logging: make it easier to emit Request-scoped logs using standard golang logging APIs

### ISSUE THREAD

**Issue by tang-fh** [googleapis/google-cloud-go#7207]
**Is your feature request related to a problem? Please describe.**

In Cloud Run, if the emitted log entry has its trace field populated with the request's trace, then those logs are grouped as children of the (Platform-emitted) request log.

The mechanism is explained in https://github.com/googleapis/google-cloud-go/blob/02c04a7c6c5e96026ce0710748a54580065a8102/logging/doc.go#L131-L148

In the case of Cloud Run, the request log has the necessary fields populated to be interpreted as a parent entry.

**Describe the solution you'd like**

I would like a method
```
func (l *Logger) RequestLogger(s Severity, r *http.Request) *log.Logger
```
analogous to 
```
func (l *Logger) StandardLogger(s Severity) *log.Logger
```
which would enable code of the form
```
type myHandler struct {
    logger *logging.Logger
}

func (h *myHandler) ServeHttp(w http.ResponseWriter, r *http.Request) {
    slg := h.logger.RequestLogger(logging.Info, r)

    slg.Println("an informative message")
}
```

**Describe alternatives you've considered**
The current workaround would be to explicitly set field `logging.Entry.Trace` with the trace value extracted from the `http.Request` object. Extracting the trace is non-trivial. See an example implementation in https://github.com/googleapis/google-cloud-go/blob/02c04a7c6c5e96026ce0710748a54580065a8102/logging/logging.go#L732-L753

**Additional context**
Logs Explorer in Cloud Console can render child log entries in tree-like format underneath the parent (request) log entries.

**Comment by daniel-sanche** [comment]
Hey Francis, thanks for putting this together!

> The current workaround would be to explicitly set field logging.Entry.Trace ...

Note that an HTTPRequest object can be attached to a log if you use the client.Logger object directly, rather than the StandardLogger. Example here: https://pkg.go.dev/cloud.google.com/go/logging#example-HTTPRequest. But after a quick look I think you're correct that there's no easy way to do this with the Standard Logger

> I would like a method
> func (l *Logger) RequestLogger(s Severity, r *http.Request) *log.Logger
> 
> analogous to
> func (l *Logger) StandardLogger(s Severity) *log.Logger

My only concern with this is that there are many other LogEntry fields that it may be useful to use through the Standard Logger (labels, resource, etc), but it doesn't seem scalable to create a different `StandardLogger()` override for each one. Maybe we should instead have something like `StandardLoggerWithExtras()` that can be expanded over time?

**Comment by minherz** [comment]
@tang-fh did you check the `HTTPRequest` structure? It has [`Request`][1] field that, if populated, is used to capture tracing information. You can ingest logs with:

```go
logger.LogSync(context.Background(), logging.Entry{
  Payload: "something happened!",
  HttpRequest: logging.HttpRequest{Request: req},
})
```

It is a little longer than your proposed API but does exactly the same thing. Using it benefits avoiding introduction of new types into the library.

I see that the populating code has a couple of potential bugs that will be great to fix though:

1. The tracing information in event it is posted using both Google tracing context and W3C tracing context is captured from Google tracing context when it should be vice versa.
1. The URL of the request is not populated although it has to be.

[1]: https://github.com/googleapis/google-cloud-go/blob/main/logging/logging.go#L475

**Comment by tang-fh** [comment]
Thanks @minherz and @daniel-sanche 

@daniel-sanche - you make a good point about other `LogEntry` fields also being of interest. As for @minherz's point, the more verbose approach which reuses the existing API might be good enough for me and side steps @daniel-sanche 's concern about fields other than trace. Let me rework some of my code which uses this cloud logging client to see how palatable I can make it.

@minherz - I can address the two potential bugs you found.

To clarify my understanding, (1) refers to the scenario where the HTTP request has both trace headers (i.e. `Traceparent` and `X-Cloud-TraceContext`). In this case, the correct behaviour is to prefer the latter (whereas the current behaviour is it takes the former).

For (2), which URL field are you referring to? I cannot find a top-level "URL" field in `LogEntry` in https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry . Are you suggesting that if we populate `trace` we also have to populate `httpRequest.requestUrl`?

**Comment by tang-fh** [comment]
I've had a tinker with some of my existing code and found a couple of issues with the alternative proposal in https://github.com/googleapis/google-cloud-go/issues/7207#issuecomment-1382360509

1. Verbosity / readability

   Using `logging.(*Logger).Log` directly makes the code less readable.

   My code changes from 
   ```
   slg.Printf("Did this")
   ...
   slg.Printf("Did that")
   ```
   to
   ```
   logger.Log(logging.Entry{
     Payload: "Did this",
     HttpRequest: logging.HttpRequest{Request: r},
   })
   ...
   logger.Log(logging.Entry{
     Payload: "Did that",
     HttpRequest: logging.HttpRequest{Request: r},
   })
   ```
   Logging now takes up significantly more of the lines and detracts from the purpose of the code itself.

2. Lost `SourceLocation`

   My initial idea for working around the verbosity issue would be to define a closure which captures `r` a la
   ```
   log := func(payload string) {
     logger.Log(logging.Entry{
       Payload: payload,
       HttpRequest: logging.HttpRequest{Request: r}, // r is captured by closure
     })
   }
   ```

   I realise that the problem with this approach is that the `SourceLocation` field will no longer be populated with the intended file/function/line data.

   To be fair, `logging.(*Logger).StandardLogger()` should have the same issue. Since `logging.(*Logger).Log` is called in `logging.severityWriter.Write()` https://github.com/googleapis/google-cloud-go/blob/6f72b3c82efb9593a548f8e8930be11d2d20931e/logging/logging.go#L310-L313 I expect all log lines will pick up `Write()` in its `SourceLocation`.

   This can be fixed by replicating the logic in `toLogEntryInternal()` https://github.com/googleapis/google-cloud-go/blob/6f72b3c82efb9593a548f8e8930be11d2d20931e/logging/logging.go#L851 I don't see a way to do this outside of the client library since the `skipLevels` argument plumbing is limited to within the library.

**Comment by tang-fh** [comment]
I've sent PR #7261 which adds a method `StandardLoggerFromEntry()` which is my interpretation of @daniel-sanche 's suggestion of `StandardLoggerWithExtras()` in comment https://github.com/googleapis/google-cloud-go/issues/7207#issuecomment-1382351320 .

The implementation is a small refactoring of `StandardLogger()` so it doesn't add any new types. Of note, it inherits the `SourceLocation` issue of `StandardLogger()` mentioned above. I can file a separate issue and fix that in another PR.

Feedback is requested.

**Comment by daniel-sanche** [comment]
> The tracing information in event it is posted using both Google tracing context and W3C tracing context is captured from Google tracing context when it should be vice versa.

I opened a new bug to track this [here](https://github.com/googleapis/google-cloud-go/issues/7264) so we don't split this discussion too much

> I've sent PR https://github.com/googleapis/google-cloud-go/pull/7261 which adds a method StandardLoggerFromEntry() which is my interpretation of @daniel-sanche 's suggestion of StandardLoggerWithExtras() in comment https://github.com/googleapis/google-cloud-go/issues/7207#issuecomment-1382351320 .
> 
> The implementation is a small refactoring of StandardLogger() so it doesn't add any new types. Of note, it inherits the SourceLocation issue of StandardLogger() mentioned above. I can file a separate issue and fix that in another PR.

Unfortunately, we are pretty strict about backwards compatibility, and it looks like both your PRs would have some breaking changes. If we are to implement this feature, we either have to make sure that existing functionality doesn't change, or stage the change for the next major version update (which I am trying to plan in a milestone [here](https://github.com/googleapis/google-cloud-go/milestone/15)). I'm thinking saving `StandardLogger` updates for the major update might make more sense, rather than trying to bandaid features on without touching existing functions

On top of that, I'm not sure if `StandardLoggerFromEntry` is a very clear, user-friendly API. It isn't clear which fields are used from the entry, and which are ignored. I think it would make more sense to use a different struct or a map to hold the config values

**Comment by tang-fh** [comment]
> I opened a new bug to track this [here](https://github.com/googleapis/google-cloud-go/issues/7264) so we don't split this discussion too much

Thanks

> > I've sent PR #7261 which adds a method StandardLoggerFromEntry() which is my interpretation of @daniel-sanche 's suggestion of StandardLoggerWithExtras() in comment [#7207 (comment)](https://github.com/googleapis/google-cloud-go/issues/7207#issuecomment-1382351320) .
> > The implementation is a small refactoring of StandardLogger() so it doesn't add any new types. Of note, it inherits the SourceLocation issue of StandardLogger() mentioned above. I can file a separate issue and fix that in another PR.
> 
> Unfortunately, we are pretty strict about backwards compatibility, and it looks like both your PRs would have some breaking changes. If we are to implement this feature, we either have to make sure that existing functionality doesn't change, or stage the change for the next major version update (which I am trying to plan in a milestone [here](https://github.com/googleapis/google-cloud-go/milestone/15)). I'm thinking saving `StandardLogger` updates for the major update might make more sense, rather than trying to bandaid features on without touching existing functions

The intention of PR #7261 wasn't to change existing functionality. It refactors struct `severityWriter` which is internal to the package and I think I have made changes so that the behaviour of `StandardLogger()` is preserved. Is there another change I missed?

> On top of that, I'm not sure if `StandardLoggerFromEntry` is a very clear, user-friendly API. It isn't clear which fields are used from the entry, and which are ignored. I think it would make more sense to use a different struct or a map to hold the config values

Maybe this can be addressed with documentation? As to which fields are used from the entry, the answer is "all of them except `Payload`". `Payload` is taken from the arguments of  the `Print()`, `Printf()`, `Println()` methods of the returned `log.Logger`.

I did wonder if using `Entry` is a form of abstraction leak but it's already used by `logging.Log()` and `logging.LogSync()`.

If you are after something that gives more explicit control over which `Entry` fields can be set (and is still extensible), I can rework the changes to use a pattern analogous to `logging.LoggerOption`. Is that what you had had in mind?

**Comment by daniel-sanche** [comment]
> The intention of PR https://github.com/googleapis/google-cloud-go/pull/7261 wasn't to change existing functionality. It refactors struct severityWriter which is internal to the package and I think I have made changes so that the behaviour of StandardLogger() is preserved. Is there another change I missed?

Yeah I think you're right, I was concerned about changing the `severityWriter` struct, but looking closer that shouldn't be a concern

> If you are after something that gives more explicit control over which Entry fields can be set (and is still extensible), I can rework the changes to use a pattern analogous to logging.LoggerOption. Is that what you had had in mind?

That is closer to what I had in mind. But there are definitely some simplicity benefits to using the entry directly as you are, since that's what is consumed by the writer later. So maybe my concerns would be better solved with careful naming and documentation.

I left another comment suggesting renaming `protoEntryWriter` to `templateEntryWriter` to avoid confusion with "protobufs". Maybe we should also rename `StandardLoggerFromEntry` to `StandardLoggerFromTemplate` or `StandardLoggerFromEntryTemplate`?

**Comment by tang-fh** [comment]
I've filed issue 7289 for the bug where `SourceLocation` is incorrect when logging via `StandardLogger()` I mentioned in comment https://github.com/googleapis/google-cloud-go/issues/7207#issuecomment-1384261216 .

I can send a PR to fix that too but it's probably better to wait for PR #7261 to be accepted first since it touches the same code.

**Comment by tang-fh** [comment]
This issue was fixed by PR #7261
